pallets/tendermint-light-client/src/lib.rs

```diff
@@ -0,0 +1,293 @@
+    };
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// ubiquitous event type
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Maximal number of finalized headers to keep in the storage, last-in first-out
+        #[pallet::constant]
+        type HeadersToKeep: Get<u32>;
+
+        /// time provider type, used to gauge whether blocks are within the trusting period
+        type TimeProvider: UnixTime;
+    }
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Pallet is halted
+        LightClientHalted,
+        /// Pallet operations are resumed        
+        LightClientResumed,
+        /// Light client is initialized
+        LightClientInitialized,
+        /// New block has been verified and imported into storage \[relayer_address, imported_block_hash\]
+        ImportedLightBlock(T::AccountId, BridgedBlockHash),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Unable to deserialize extrinsic
+        DeserializeError,
+        /// Light client has not been initialized        
+        NotInitialized,
+        /// Light client has already been initialized
+        AlreadyInitialized,
+        /// Light client is currently halted
+        Halted,
+        /// The minimum voting power threshold is not reached, the block cannot be trusted yet
+        NotEnoughTrust,
+        /// Verification failed, the block is invalid.        
+        InvalidBlock,
+    }
+
+    // NOTE for storage:
+    // - header storage should be a ring buffer (i.e. we keep n last headers, ordered by the insertion time)
+    // - we keep a pointer to the last finalized header (to avoid deserializing the whole buffer)
+    // - insertion moves the pointer and updates the buffer
+    // or?
+    // https://substrate.recipes/ringbuffer.html
+
+    /// Hash of the last imported header from the bridged chain
+    #[pallet::storage]
+    #[pallet::getter(fn get_last_imported_hash)]
+    pub type LastImportedHash<T: Config> = StorageValue<_, BridgedBlockHash, ValueQuery>;
+
+    /// A ring buffer of imported hashes "ordered" by their insertion time
+    #[pallet::storage]
+    pub type ImportedHashes<T: Config> = StorageMap<_, Identity, u32, BridgedBlockHash>;
+
+    /// Current ring buffer position
+    #[pallet::storage]
+    pub(super) type ImportedHashesPointer<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+    /// Bridged chain Headers which have been imported by the client
+    /// Client keeps HeadersToKeep number of these at any time
+    #[pallet::storage]
+    pub(super) type ImportedBlocks<T: Config> =
+        StorageMap<_, Identity, BridgedBlockHash, LightBlockStorage, OptionQuery>;
+
+    // TODO : expose in runtime API and nodes RPC
```

Comment:
YMMV, but sometimes it just *feels* right to leave a TODO comment as a constant reminder for self (I look at it way more often than at Jira) - and definitely beats not leaving any comment at all and forgetting. On the other hand in this case it will eventually surface, as other pallets in the bridge ecosystem will eventually need the query, so I opted for just removing the comment.